### ISSUE TITLE
GraphQL schema parsing errors

### ISSUE THREAD

**Issue by ivarref** [walmartlabs/lacinia#259]
Hi, and thanks for a fine library!

I believe there are several related errors in parsing GraphQL schemas. Two examples are given below. The EDN equivalents both work as expected.

```clojure
(ns lacinia-parse-errors.core ; setup ...
  (:require [com.walmartlabs.lacinia.parser.schema :as parse-schema]
            [com.walmartlabs.lacinia.schema :as schema]
            [com.walmartlabs.lacinia :refer [execute]]))

; The following fails:
(comment
  (parse-schema/parse-schema
    "type QueryRoot {
      hello(input: String): String
      #     ^^^^^ Here is the problem.
      #           It seems that 'input' is some kind of reserved keyword (or possibly a separate lexer token?)
      #           Changing 'input' to for example 'inp' solves the problem.
    }"
    {:resolvers {:QueryRoot {:hello (constantly "Hello World")}}}))
; CompilerException clojure.lang.ExceptionInfo: Failed to parse GraphQL schema. 
; {:errors ({:locations [{:line 2, :column nil}], :message "mismatched input 'input' 
; expecting {Name, StringValue, BlockStringValue}"} 
; {:locations [{:line 2, :column nil}], :message "mismatched input ')' 
; expecting {'}', '@', '!', Name, StringValue, BlockStringValue}"} 
; {:locations [{:line 3, :column nil}], :message "mismatched input '}' 
; expecting {':', '('}"})}, compiling:(/home/ire/code/learn/lacinia-parse-errors/src/lacinia_parse_errors/core.clj:8:3) 


; The EDN equivalent works:
(comment
  (-> {:queries
       {:hello {:type    'String
                :args    {:input {:type 'String}}
                :resolve (fn [context args value]
                           (str "Hello " (get args :input "World")))}}}
      (schema/compile)
      (execute "{ hello(input:\"Bar\") }" nil nil)))


; The following fails:
(comment
  (parse-schema/parse-schema
    "type X {
       type: String
      #^^^^ Similarly: Here is the problem.
      #     Changing 'type' to 'typ' solves the problem.
     }"
    {}))
; CompilerException clojure.lang.ExceptionInfo: Failed to parse GraphQL schema. 
; {:errors ({:locations [{:line 2, :column nil}], :message "mismatched input 'type' 
; expecting {Name, StringValue, BlockStringValue}"} 
; {:locations [{:line 2, :column nil}], :message "extraneous input ':' expecting Name"} 
; {:locations [{:line 3, :column nil}], :message "mismatched input '}' 
; expecting {'{', '@', 'implements'}"})}, compiling:(/home/ire/code/learn/lacinia-parse-errors/src/lacinia_parse_errors/core.clj:27:3) 


; The EDN equivalent works:
(comment
  (-> {:objects
       {:X {:fields {:type {:type 'String}}}}
       :queries
       {:X {:type    '(non-null :X)
            :args    {}
            :resolve (constantly {:type "foo"})}}}
      (schema/compile)
      (execute "{ X { type } }" nil nil)))
```

I guess the ANTLR4 grammar is somehow to blame, but I don't know enough to fix it.

PS: The [alumbra.parser](https://github.com/alumbra/alumbra.parser) parses these documents fine, and also uses ANTLR4.

I'm using lacinia 0.32.0-rc-2 and Clojure 1.9.0.

Thanks and kind regards.

**Comment by hlship** [comment]
What version of Lacinia are you using?  An earlier version mistakenly required a `{` ... `}` around the schema document.

**Comment by ivarref** [comment]
I’m using the latest version, 0.32.0-rc-2.

**Comment by ivarref** [comment]
I've updated the issue with a more detailed elaboration about what is failing @hlship

**Comment by ivarref** [comment]
For
```graphql
type QueryRoot {
  hello(input: String): String
}
```

the following changes to `schema.g4` solved the problem:

```antlr
inputTypeDef
  : description? K_INPUT Name directiveList? fieldDefs
  ;

...

Name
  : [_A-Za-z][_0-9A-Za-z]*
  ;

K_INPUT
    : 'input'
    ;
...
```

If placing `K_INPUT` above `Name`, the parse error persists, so I suppose the default behaviour is to extract parser literals and place them in priority above stated lexer rules. Please correct me if I'm wrong.

Would a pull request moving parser rules literals below `Name` (and possibly other rules too) be of interest?

**Comment by hlship** [comment]
If you have time for a PR and to sign the CLA that would be great. Otherwise I can look into this along with proper parsing of implemented interfaces.

We've hit this kind of problem repeatedly with Antlr ... a keyword has to be a keyword everywhere, whereas the parser used in JavaScript impl can be more forgiving.  But you have the basic solution, controlling the matching order, and I recently refactored the parsing code (into two steps) that makes it dramatically easier to test and maintain these special cases.

**Comment by ivarref** [comment]
I've added a pull request and signed the CLA @hlship .

https://github.com/walmartlabs/lacinia/pull/263

Thanks and kind regards.

**Comment by hlship** [comment]
This has been fixed in 0.33.0 (I just checked your examples at the REPL), and there are already a selection of issues covering thes cases. Thanks for the report.
